cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,289 @@
+	// Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	//
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec contains specification of the ProvisioningRequest object.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	//
+	// +kubebuilder:validation:Required
+	Spec ProvisioningRequestSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+	// Status of the ProvisioningRequest. CA constantly reconciles this field.
+	//
+	// +optional
+	Status ProvisioningRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ProvisioningRequestList is a object for list of ProvisioningRequest.
+type ProvisioningRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	//
+	// +optional
+	metav1.ListMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+	// Items, list of ProvisioningRequest returned from API.
+	//
+	// +optional
+	Items []ProvisioningRequest `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ProvisioningRequestSpec is a specification of additional pods for which we
+// would like to provision additional resources in the cluster.
+type ProvisioningRequestSpec struct {
+	// PodSets lists groups of pods for which we would like to provision
+	// resources.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	PodSets []PodSet `json:"podSets" protobuf:"bytes,1,rep,name=podSets"`
+
+	// ProvisioningClass describes the different modes of provisioning the resources.
+	// Supported values:
+	// * kubernetes.io/checkCapacity - check if current cluster state can fullfil this request, do not reserve the capacity.
+	// * kubernetes.io/atomicScaleUp - provision the resources in an atomic manner
+    // * ... - potential other classes that are specific to the cloud providers
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	ProvisioningClass string `json:"provisioningClass" protobuf:"bytes,2,name=provisioningClass"`
+
+	// AdditionalParameters contains all other parameters custom classes may require.
+	//
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	AdditionalParameters map[string]string `json:"additionalParameters" protobuf:"bytes,3,name=additionalParameters"`
+}
+
+type PodSet struct {
+	// TemplateRef is a reference to a PodTemplate object that is representing pods 
+	// that will consume this reservation. Users need to make sure that the
+	// fields relevant to scheduler (e.g. node selector tolerations) are
+	// consistent between this template and actual pods consuming the Provisioning Request.
+	TemplateRef *metav1.ObjectReference `json:"templateRef" protobuf:"bytes,1,name=templateRef"`
```

Comment:
This requires the the pods exist in order to piggy-back off of their existing specs. If I'm trying to deploy both this resource and something like a Deployment or ReplicaSet object at the same time, I won't know what the pods names/namespaces are before the relevant controllers actually deploy them. Does it make more sense to have this reference higher-level objects that control pod deployments but have pod template fields or to have the podSpecs statically written into the pod set?

Reply:
> Does it make more sense to have this reference higher-level objects that control pod deployments but have pod template fields or to have the podSpecs statically written into the pod set?

Fro the CUJ mentioned above it makes sense, but imposes a serious debt on the CA code-base as either it has to be able to access all K8s higher level objects (this will not play well with custom resources) or have a complex dynamic client to access a given field of an object (which may not necessary be a valid pod template).

Due to the above the approaches that make most sense are to embed the whole template or reference the pod template object. The later on being the proffered by the @liggitt